WIP: Add item virtualization

grokys opened the pull request:
This PR adds item virtualization to Avalonia. The main changes are as follows:
- Adds an `IVirtualizingPanel` interface
- Adds a `VirtualizingStackPanel` control
- Adds a `VirtualizationMode` to `ListBox`
- Adds support for virtualization to `ItemsPresenter` and `ItemContainerGenerator`

Breaking changes
- `IScrollable` renamed to `ILogicalScrollable`
- `ItemContainerGenerator.Materialize` now works on single items instead of creating batches

Currently supported virtualization modes are:
- `None` - no virtualization, as previously
- `Simple` - "coarse" scrolling by item as in WPF

The basic flow is:
- `ListBox` has a `VirtualizingStackPanel` as its `ItemsPanel`
- `ItemsPresenter` notices this
- When `ItemsPresenter` is arranged it uses the `IVirtualizingPanel` interface to fill the panel with items. Once the panel is full it stops.
- `ItemsPresenter` implements `ILogicalScrollable`
- When the `ItemsPresenter`'s offset changes, it recycles the containers with help from `ItemContainerGenerator` to display different items (and changes the order of the items if the scroll is by less than a page - i.e. scrolling by half a page moves the bottom half of the items to the top, and recycles the top half of the items and moves them to the bottom)

Still to implement:
- [x] Partial items at the bottom are obscured
- [x] When at the bottom of the list and enlarged, items should be added to the top
- [x] Support INCC
- [x] Keyboard support for moving to as-yet-unmaterialized items with arrow keys
- [x] Use a "sparse" array for containers in `ItemContainerGenerator`
- [x] Expand `ItemsControl` to fill available space
- [x] Handle null values in `Items`
- [x] Recycle `DataTemplate`s
- [ ] ~~"Smooth" scrolling - we might need need two modes:~~ going to punt on this until later
  - [ ] ~~An "average" mode for lists where the items are uniform size~~
  - [ ] ~~"advanced" mode where all the items have to be measured somehow (not sure sure about this one)~~

grokys committed:
WIP

grokys committed:
WIP: More work on primitive virtualization.

grokys committed:
Removed IScrollable from VirtualizingStackPanel

grokys committed:
Added pixel offset to virtualizing panel.

grokys committed:
Leave one container materialized.

grokys committed:
Fix scrolling algorithm

grokys committed:
Added IScrollable.IsLogicalScrollEnabled.

grokys committed:
Merge branch 'master' into item-virtualization-avalonia

grokys committed:
WIP: Adding virtualization to ItemsPresenter.

grokys committed:
More WIP on adding virtualization to ItemsPresenter.

grokys committed:
Dematerialize items when removed.

grokys committed:
Refactored virtualization handling into classes

Also take into account the scroll direction of the panel.

grokys committed:
WIP Trying to get virt scrolling working.

grokys committed:
Ensure to update scroll info when items changed.

grokys committed:
More WIP on virtualization.

Kinda nearly working in the test app.

grokys committed:
Update scroll after arrange.

grokys committed:
Handle scrolling >1 page.

grokys committed:
Removed Thingamybob.

grokys committed:
Delegate ScrollIntoView to ItemVirtualizer.

This doesn't make sense for ItemVirtualizationMode.Simple so we do
nothing here.

grokys committed:
Renamed IScrollable -> ILogicalScrollable

We're going to need IScrollable as an interface to ScrollViewer-type
controls.

grokys committed:
Added IScrollable interface.

This is now the basic Extent/Offset/Viewport interface. Implemented by
ScrollViewer and ScrollContentPresenter.

grokys committed:
Expose IScrollable in ListBox.

To allow programmatic control of the ListBox's scroll.

grokys committed:
Simplify ListBox unit test template creation.

grokys committed:
Update ListBoxTests

Now ListBox is virtualized it needs more setup for unit testing.

grokys committed:
Fix virtualized item selection.

So that recycled items' selection state is set correctly.

grokys committed:
Don't try to dematerialize null containers.

grokys committed:
Added ListBox.VirtualizationMode

grokys committed:
Added virtualization test app.

grokys committed:
Show scroll info in virt test app.

grokys committed:
Make scrolling up >= 1 page work

grokys committed:
Update containers when Items changes.

grokys committed:
Clarified variable name.

grokys committed:
Added "Recreate" button to virtualization test app.

grokys committed:
Correctly resize items.

mterwoord commented:
""ItemContainerGenerator.Materialize now works on single items instead of creating batches"
Why not keep it doing batches: What if you suddenly need 10 more?

grokys commented:
@mterwoord this is because when you're creating items for a virtualized panel, you don't know how many items will fit until each item has been measured, so you have to create them one-by-one.

In the non-virtualized case, this just moves the loop outside of `ItemContainerGenerator`.

mterwoord commented:
True!

grokys committed:
Dematerialize correct items.

grokys committed:
Added failing partial item tests.

grokys committed:
Correctly handle partially obscured items.

And move logical for selecting horizontal/vertical components of
extent/offset/viewport into `ItemVirtualizer` base class.

Commit:
Split virtualization tests in two

One for base virtualization tests and one for simple mode tests.

grokys committed:
Add containers at beginning when scrolled to end.

grokys committed:
WIP: INCC virtualization support.

Commit:
added support for add items and remove items in virtualization sample app.

Commit:
added button to prepend and append items to list.

Commit:
add items at selected index, or at end if no selection.

Commit:
add items, adds after selected item.

Commit:
Added test for removing last item in the list, when scrolled to end.

Commit:
Throw not implemented exception placeholder.

Commit:
added some more test cases for simple virtualization.

Commit:
added implementation for removing first and last elements.

grokys committed:
Merge pull request #554 from VitalElement/grokys/item-virtualization-avalonia

WIP Further Implementation of INCC handling

grokys committed:
More work on virtualization with INCC.

Got removes working a bit better.

grokys committed:
Allow multiple selection in virt test app.

grokys committed:
More work on items removed.

grokys committed:
Make inserting items more user friendly.

grokys committed:
Tweaked comments.

grokys committed:
Handle move and replace in virtualized lists.

grokys committed:
Fix CarouselPresenter.

Forgot to assign to return value!

grokys committed:
Added documentation for item virtualizers.

grokys committed:
Use dictionary to track item containers.

Because even when virtualized we were still creating a list the size of
the Items collection to store the containers. Using Dictionary here
still isn't ideal - we'd ideally use some sort of sparse array but that
can be optimized later.

grokys committed:
Make ItemContainerGenerator create containers

Make the untyped ItemContainerGenerator create `ContentPresenter`
containers when the data isn't a control. This is required for correct
binding to `DataContext` in a data template.

grokys committed:
Tweaked a couple of tests.

grokys committed:
Recycle DataTemplates.

If a `ContentPresenter`s content is assigned a value which matches the
current `DataTemplate` then just change the
`ContentPresenter.DataContext` and the existing item will update itself.

grokys committed:
Fixed DevTools.

- Set `InheritanceParent` before Parent so that control doesn't
initially inherit the wrong DataContext
- Don't set `InheritanceParent` when setting `Parent` if
`InheritanceParent` already set
- Added `IDataTemplate.SupportsRecycling`

grokys committed:
Added Move and MoveRange to AvaloniaList.

Also added test for replace operation.

grokys committed:
Move items in panel Children collection

Instead of removing and re-adding them as this causes a removal and
re-addition to the logical tree, with all the overhead that involves.

grokys committed:
Set ContentPresenter.DataContext in UpdateChild.

Thus delaying any changes that changing the data context would make
until we know we have the correct child.

grokys committed:
Allow virtualized list to take up available space.

grokys committed:
Fix virtualized list to taking up available space.

Previous commit was slightly wrong: too many controls were being created
for the list in VirtualizingTest.

grokys committed:
Initial impl. of scrolling with arrow keys.

- Currently only vertical implemented
- Doesn't handle partially visible items at end of list

grokys committed:
Updated overflow algorithm.

grokys committed:
Use a full item's offset for overflow.

grokys committed:
Fix keyboard nav with partially visible items.

Commit:
Support horiz keyboard movement.

Commit:
Fix compile error.

Commit:
Allow changing virtualization mode in test app.

Not yet working correctly.

Commit:
Fix PixelOverflow during arrange.

Commit:
Fix changing virtualization mode in theory.

That is; tests pass. In practise in the virtualization test app it still
doesn't quite work.

Commit:
Fixed unit test name.

Commit:
Fixed ScrollContentPresenter.

Did not react correctly to child changing.

Commit:
Tweak measure/arrange algorithm.

To ensure that parent controls are measured/arranged before children.

Commit:
Tidied ScrollContentPresenter logic.

Commit:
Added failing test.

Commit:
Fix copy/paste error in LayoutManager.

Fixes test previously added.

Commit:
Don't create items immediately...

...in virtualized ItemsPresenters.

Commit:
FocusNavigationDirection => NavigationDirection

And added page up/down.

Commit:
Added page up/down support to virt lists.

Commit:
Use correct ItemContainerGenerator for menus.

Commit:
Allow changing orientation in virt sample.

Commit:
Fixed scrolling differently sized items.

Commit:
Fix failing tests.

Commit:
Implement AutoScrollToSelectedItem

For virtualizing lists.

Commit:
Implement Home and End navigation.

Commit:
Fix directional nav for non-virtualized lists.

Except Page Up/Down - these will need some extra code.

Commit:
Fix menu separators.

grokys committed:
Added failing binding tests.

grokys committed:
Revert "Set ContentPresenter.DataContext in UpdateChild."

This reverts commit f3c7ea27a1e98dbcd777e27b9bc2a5eb1fffbdff.

grokys committed:
Removed test.

Decided that we shouldn't do this.

grokys committed:
Make broken DataContext bindings produce null.

This prevents incorrect DataContexts cascading down to children when the
DataContext binding is invalid, e.g. when things are being set up.

grokys committed:
Don't delay bind to DataContext.

DataContext bindings are special in that they return null instead of
UnsetValue when broken, and it's important that they do that when things
are being set up.

grokys committed:
Merge branch 'binding-refactor' into item-virtualization-avalonia

grokys committed:
Revert "Revert "Set ContentPresenter.DataContext in UpdateChild.""

This reverts commit bf02ae41a9050a2e2db21dadb9970d9248424f99.
